### PR TITLE
Feat marxan 1055 dump single blm result

### PR DESCRIPTION
**Pull request by aciddaute**
### Overview

_Please write a description. If the PR is hard to understand, provide a quick explanation of the code._

### Designs

_Link to the related design prototypes (if applicable)_

### Testing instructions

_Please explain how to test the PR: ID of a dataset, steps to reach the feature, etc._

### Feature relevant tickets

_Link to the related task manager tickets_

---

## Checklist before submitting

- [ ] Meaningful commits and code rebased on `develop`.
- [ ] If this PR adds feature that should be tested for regressions when
      deploying to staging/production, please add brief testing instructions
      to the deploy checklist (`docs/deployment-checklist.md`)
- [ ] Update CHANGELOG file

### CODE REVIEW

File: api/apps/geoprocessing/src/marxan-sandboxed-runner/adapters-blm/blm-run-adapter.module.ts [hunk 1]
```diff
@@ -20,11 +20,19 @@ import { MostDifferentService } from '@marxan-geoprocessing/marxan-sandboxed-run
 import { BestSolutionService } from '@marxan-geoprocessing/marxan-sandboxed-runner/adapters-single/solutions-output/best-solution.service';
 import { MarxanDirectory } from '@marxan-geoprocessing/marxan-sandboxed-runner/adapters-single/marxan-directory.service';
 import { BlmResultsParser } from '@marxan-geoprocessing/marxan-sandboxed-runner/adapters-blm/blm-results.parser';
+import { BlmPartialResultsRepositoryModule } from '@marxan-geoprocessing/marxan-sandboxed-runner/adapters-blm/repositories/partial/blm-partial-results-repository.module';
+import { BlmFinalResultsRepositoryModule } from '@marxan-geoprocessing/marxan-sandboxed-runner/adapters-blm/repositories/final/blm-final-results-repository.module';
```
Comment (kgajowy) [inline api/apps/geoprocessing/src/marxan-sandboxed-runner/adapters-blm/blm-run-adapter.module.ts:24]: relative imports.

File: api/apps/geoprocessing/src/marxan-sandboxed-runner/adapters-blm/marxan-sandbox-blm-runner.service.ts [hunk 2]
```diff
@@ -80,6 +88,18 @@ export class MarxanSandboxBlmRunnerService
...
+              output,
+              forScenarioId,
+              blmValue,
+            );
+            await this.blmFinalRepository.saveBest(
+              output,
+              calibrationId,
+              forScenarioId,
+              blmValue,
+            );
```
Comment (kgajowy) [inline api/apps/geoprocessing/src/marxan-sandboxed-runner/adapters-blm/marxan-sandbox-blm-runner.service.ts:101]: [blocker] Please read the .md docs. We shouldn't save anything in `final` repository until all partial results are gathered.

File: api/libs/blm-calibration/src/blm-partial-results.geo.entity.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
+import { Column, Entity, PrimaryGeneratedColumn } from 'typeorm';
+
+const tableName = `blm_partial_results`;
+
+@Entity(tableName)
+export class BlmPartialResultEntity {
```
Comment (kgajowy) [inline api/libs/blm-calibration/src/blm-partial-results.geo.entity.ts:6]: This entity is implementation details to calibration runner and do not have to be exposed in `lib`

File: api/apps/geoprocessing/src/marxan-sandboxed-runner/adapters-blm/repositories/final/blm-final-results.repository.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import { ExecutionResult } from '@marxan/marxan-output';
+
+export abstract class BlmFinalResultsRepository {
```
Comment (kgajowy) [inline api/apps/geoprocessing/src/marxan-sandboxed-runner/adapters-blm/repositories/final/blm-final-results.repository.ts:3]: [blocker] seems to be a port but is within adapters? What happened to original ports?

**Review by kgajowy — CHANGES_REQUESTED** [review]

File: api/apps/geoprocessing/src/marxan-sandboxed-runner/adapters-blm/blm-final-results.repository.ts [hunk 5]
```diff
@@ -1,35 +1,49 @@
...
+
+      await txManager.save(
+        BlmFinalResultEntity,
+        partialResults.map((partial) => ({
+          scenarioId: partial.scenarioId,
+          blmValue: partial.blmValue,
+          score: partial.score,
+          boundaryLength: partial.boundaryLength,
+        })),
+      );
```
Comment (kgajowy) [inline api/apps/geoprocessing/src/marxan-sandboxed-runner/adapters-blm/blm-final-results.repository.ts:42]: Shouldn't be a big deal as we won't have much rows there but consider using raw SQL:

```
insert into final_results fr (scenario_id, blm_value, score, boundary_length)
select scenario_id, blm_value, score, boundary_length) from partial_results pr 
where pr.scenario_id = $1
```

File: api/apps/geoprocessing/src/marxan-sandboxed-runner/adapters-blm/blm-partial-results.repository.ts [hunk 6]
```diff
@@ -1,37 +1,61 @@
...
+    const runs = this.bestSolutionsService.map(
+      this.marxanOutputParser.parse(runsSummary),
+    );
+    const bestRun = runs.find((run) => run.best);
+
+    if (!bestRun) {
+      throw new Error(
+        `${scenarioId} scenario calibration doesn't have a best run for ${blmValue} BLM value`,
+      );
+    }
```
Comment (kgajowy) [inline api/apps/geoprocessing/src/marxan-sandboxed-runner/adapters-blm/blm-partial-results.repository.ts:52]: This part could be a standalone class to make it easier to test and separate concerns.

File: api/apps/geoprocessing/src/marxan-sandboxed-runner/adapters-blm/blm-partial-results.repository.ts [hunk 7]
```diff
@@ -1,37 +1,61 @@
...
+
+    await this.repository.save({
+      blmValue,
+      boundaryLength: bestRun.connectivity,
+      scenarioId,
+      score: bestRun.score,
+    });
```
Comment (kgajowy) [inline api/apps/geoprocessing/src/marxan-sandboxed-runner/adapters-blm/blm-partial-results.repository.ts:59]: Consider the scenario:

1. Run Calibration for [0.1-1.0] range (fails at the 6th blm value
2. Run Calibration for [2-10] range

Boom, partial results (and final!) contains 11 values - as `move` to `final` repository didn't have a chance to run and thus to clear partials.

Possible option is to put an arbitrary id of `calibration` that we values as inserted - and remove all values related to given scenario that are not using that arbitrary ID. (or just remove all partials

File: api/apps/geoprocessing/src/marxan-sandboxed-runner/adapters-blm/marxan-sandbox-blm-runner.service.ts [hunk 8]
```diff
@@ -36,26 +34,11 @@ export class MarxanSandboxBlmRunnerService
...
-
-      const singleRunner = new MarxanSandboxRunnerService(
-        workspaceBuilder,
-        inputFilesHandler,
-        outputHandler,
+    for (const { workspace, blmValue } of workspaces) {
+      const singleRunner = this.marxanRunnerFactory.for(
+        scenarioId,
+        blmValue,
+        workspace,
```
Comment (kgajowy) [inline api/apps/geoprocessing/src/marxan-sandboxed-runner/adapters-blm/marxan-sandbox-blm-runner.service.ts:41]: Really like it now :D 🚀

File: api/apps/geoprocessing/src/marxan-sandboxed-runner/adapters-blm/blm-final-results.repository.ts (hunk 5 above)
Comment (aciddaute) [reply api/apps/geoprocessing/src/marxan-sandboxed-runner/adapters-blm/blm-final-results.repository.ts:42]: Ok, I'll change it :+1:

File: api/apps/geoprocessing/src/marxan-sandboxed-runner/adapters-blm/blm-partial-results.repository.ts (hunk 6 above)
Comment (aciddaute) [reply api/apps/geoprocessing/src/marxan-sandboxed-runner/adapters-blm/blm-partial-results.repository.ts:52]: Maybe I can create `BestRunService`, what do you think?

File: api/apps/geoprocessing/src/marxan-sandboxed-runner/adapters-blm/marxan-sandbox-blm-runner.service.ts (hunk 8 above)
Comment (aciddaute) [reply api/apps/geoprocessing/src/marxan-sandboxed-runner/adapters-blm/marxan-sandbox-blm-runner.service.ts:41]: 🚀 👍

File: api/apps/geoprocessing/src/marxan-sandboxed-runner/adapters-blm/blm-partial-results.repository.ts (hunk 7 above)
Comment (aciddaute) [reply api/apps/geoprocessing/src/marxan-sandboxed-runner/adapters-blm/blm-partial-results.repository.ts:59]: But, when should be removed partial results with a different `calibrationId`? On each iteration? Maybe I should try catching a possible error and then removing all partial results for that specific scenario? It sound good to you?

File: api/apps/geoprocessing/src/marxan-sandboxed-runner/adapters-blm/blm-partial-results.repository.ts (hunk 6 above)
Comment (kgajowy) [reply api/apps/geoprocessing/src/marxan-sandboxed-runner/adapters-blm/blm-partial-results.repository.ts:52]: @aciddaute 👍 (be aware of the class names clashing tho - NestJS DI system uses class names)

File: api/apps/geoprocessing/src/marxan-sandboxed-runner/adapters-blm/blm-partial-results.repository.ts (hunk 7 above)
Comment (kgajowy) [reply api/apps/geoprocessing/src/marxan-sandboxed-runner/adapters-blm/blm-partial-results.repository.ts:59]: assuming we put [scenarioId, calibrationId], we can try to remove "`where: {scenarioId, calibrationId: Not(...)}` or something like that. Isn't perfect, so feel free to find out. Not sure if removing "partials" upon a fail is fine - might be this is expected for some cases and partial results are fine - we don't know yet ;)

**Review by kgajowy — APPROVED** [review]

**Pull request merged**